cmd/agent/app/builder.go

```diff
@@ -152,86 +146,46 @@ type SamplingServerConfiguration struct {
 	HostPort string `yaml:"hostPort" validate:"nonzero"`
 }
 
-// WithReporter adds auxiliary reporters
+// WithReporter adds auxiliary reporters.
 func (b *Builder) WithReporter(r reporter.Reporter) *Builder {
 	b.otherReporters = append(b.otherReporters, r)
 	return b
 }
 
-// WithDiscoverer sets service discovery
-func (b *Builder) WithDiscoverer(d discovery.Discoverer) *Builder {
-	b.discoverer = d
-	return b
-}
-
-// WithDiscoveryNotifier sets service discovery notifier
-func (b *Builder) WithDiscoveryNotifier(n discovery.Notifier) *Builder {
-	b.notifier = n
-	return b
-}
-
-// WithCollectorServiceName sets collector service name
-func (b *Builder) WithCollectorServiceName(s string) *Builder {
-	b.CollectorServiceName = s
-	return b
-}
-
-// WithChannel sets tchannel channel
-func (b *Builder) WithChannel(c *tchannel.Channel) *Builder {
-	b.channel = c
-	return b
-}
-
-func (b *Builder) enableDiscovery(channel *tchannel.Channel, logger *zap.Logger) (interface{}, error) {
-	if b.discoverer == nil && b.notifier == nil {
-		return nil, nil
+func (b *Builder) createMainReporter(mFactory metrics.Factory, logger *zap.Logger) (*tchreporter.Reporter, error) {
+	if len(b.Builder.CollectorHostPorts) == 0 {
+		b.Builder.CollectorHostPorts = b.CollectorHostPorts
 	}
-	if b.discoverer == nil || b.notifier == nil {
-		return nil, errors.New("both discovery.Discoverer and discovery.Notifier must be specified")
+	if b.Builder.CollectorServiceName == "" {
+		b.Builder.CollectorServiceName = b.CollectorServiceName
 	}
-
-	logger.Info("Enabling service discovery", zap.String("service", b.CollectorServiceName))
-
-	subCh := channel.GetSubChannel(b.CollectorServiceName, tchannel.Isolated)
-	peers := subCh.Peers()
-	return peerlistmgr.New(peers, b.discoverer, b.notifier,
-		peerlistmgr.Options.MinPeers(defaultInt(b.DiscoveryMinPeers, defaultMinPeers)),
-		peerlistmgr.Options.Logger(logger))
+	if b.Builder.DiscoveryMinPeers == 0 {
+		b.Builder.DiscoveryMinPeers = b.DiscoveryMinPeers
+	}
+	return b.Builder.CreateReporter(mFactory, logger)
 }
 
 // CreateAgent creates the Agent
 func (b *Builder) CreateAgent(mFactory metrics.Factory, logger *zap.Logger) (*Agent, error) {
-	if b.channel == nil {
-		// ignore errors since it only happens on empty service name
-		b.channel, _ = tchannel.NewChannel(agentServiceName, nil)
-	}
-
-	if b.CollectorServiceName == "" {
-		b.CollectorServiceName = defaultCollectorServiceName
-	}
-
-	// Use static collectors if specified.
-	if len(b.CollectorHostPorts) != 0 {
-		d := discovery.FixedDiscoverer(b.CollectorHostPorts)
-		b = b.WithDiscoverer(d).WithDiscoveryNotifier(&discovery.Dispatcher{})
-	}
-
-	discoveryMgr, err := b.enableDiscovery(b.channel, logger)
+	mainReporter, err := b.createMainReporter(mFactory, logger)
```

Comment:
I think this is overly restrictive. 
In the sense, why does one need to launch an agent with the main reporter?
What if I want to launch an agent with only a custom reporter?